pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -103,48 +115,193 @@ type controllerVolumeTopology struct {
+				if err != nil {
+					log.Errorf("failed to create an informer for CSINodeTopology instances. Error: %+v", err)
+					return nil, err
+				}
+				// If AvailabilityZone CR not registered, no need to initiate the informer.
+				if azInformer == nil {
+					log.Info("Initiating AvailabilityZone informer not required.")
+					return nil, nil
+				}
+				wcpControllerVolumeTopologyInstance = &wcpControllerVolumeTopology{
+					k8sConfig:  config,
+					azInformer: *azInformer,
+				}
 			}
-			log.Info("Topology service initiated successfully")
+		} else {
+			controllerVolumeTopologyInstanceLock.RUnlock()
 		}
-	} else {
-		controllerVolumeTopologyInstanceLock.RUnlock()
+		return wcpControllerVolumeTopologyInstance, nil
 	}
-	return controllerVolumeTopologyInstance, nil
+	return nil, logger.LogNewErrorf(log, "InitTopologyServiceInController not implemented for "+
+		"cluster flavor: %q", c.clusterFlavor)
+}
+
+// startAvailabilityZoneInformer listens on changes to AvailabilityZone instances and updates the azClusterMap cache.
+func startAvailabilityZoneInformer(ctx context.Context, cfg *restclient.Config) (*cache.SharedIndexInformer, error) {
+	log := logger.GetLogger(ctx)
+	// Create an informer for AvailabilityZone instances.
+	dynInformer, err := k8s.GetDynamicInformer(ctx, "topology.tanzu.vmware.com",
+		"v1alpha1", "availabilityzones", metav1.NamespaceAll, cfg, true)
+	if err != nil {
+		// TODO: TKGS-HA - Verify if this works
+		if apiMeta.IsNoMatchError(err) {
+			log.Info("AvailabilityZone CR is not registered on the cluster")
+			return nil, nil
+		}
+		log.Errorf("failed to create dynamic informer for AvailabilityZone CR. Error: %+v",
+			err)
+		return nil, err
+	}
+	availabilityZoneInformer := dynInformer.Informer()
+	availabilityZoneInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			azCRAdded(obj)
+		},
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			azCRUpdated(oldObj, newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			azCRDeleted(obj)
+		},
+	})
+
+	// Start informer.
+	go func() {
+		log.Info("Informer to watch on AvailabilityZone CR starting..")
+		availabilityZoneInformer.Run(make(chan struct{}))
+	}()
+	return &availabilityZoneInformer, nil
+}
+
+// azCRAdded handles adding AZ name and clusterMoref to the cache.
+func azCRAdded(obj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Retrieve name of CR instance.
+	azName, found, err := unstructured.NestedString(obj.(*unstructured.Unstructured).Object, "metadata", "name")
+	if !found || err != nil {
+		log.Errorf("failed to get `name` from AvailabilityZone instance: %+v, Error: %+v", obj, err)
+		return
+	}
+	// Retrieve clusterMoref from instance spec.
+	// TODO: TKGS-HA - convert to slice when appropriate
+	clusterComputeResourceMoId, found, err := unstructured.NestedString(obj.(*unstructured.Unstructured).Object,
+		"spec", "clusterComputeResourceMoId")
+	if !found || err != nil {
```

Comment:
I think we should fail the container, as this is very critical information we missed to cache during CR addition.
here we are just logging the error and continuing

Reply:
Talked to Divyen and decided to place checks with proper errors in CreateVolume workflow where this cache is used.